### PR TITLE
Add config to docs

### PR DESCRIPTION
**Pull request by allevo**
Add `reply.context.config` to documentation.
Add typescript definition too

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by jsumners** [comment]
Isn't this already documented at https://github.com/fastify/fastify/blob/5fefda9dd2f882e807b7ae736d39be4860c13d6b/docs/Plugins.md#plugin-options ?

File: docs/Routes.md [hunk 1]
```diff
@@ -230,3 +230,24 @@ fastify.get('/', { logLevel: 'warn' }, (request, reply) => {
...
+
+```js
+// server.js
+const fastify = require('fastify')()
+
+function handler (req, reply) {
+  reply.send(reply.context.config.output)
+}
+
+fastify.register('/en', { config: { output: 'hello world!' } }, handler)
```
Comment (delvedor) [inline docs/Routes.md]: I guess this should be a `get`.

File: docs/Routes.md [hunk 2]
```diff
@@ -230,3 +230,24 @@ fastify.get('/', { logLevel: 'warn' }, (request, reply) => {
...
+fastify.register('/it', { config: { output: 'ciao mondo!' } }, handler)
+
+fastify.listen(3000)
+```
+
+*Remember to not abuse of this feature. Please consider to use `decorate`, `decorateRequest` and `decorateReply` methods before*
```
Comment (delvedor) [inline docs/Routes.md]: Why?

**Comment by jsumners** [comment]
I think the two sections of the documentation should reference each other.

File: fastify.d.ts [hunk 3]
```diff
@@ -69,6 +69,7 @@ declare namespace fastify {
     send: (payload?: any) => FastifyReply<HttpResponse>
     sent: boolean
     res: HttpResponse
+    context: any
```
Comment (mcollina) [inline fastify.d.ts]: this should be something more specific. Moreover, it does not define `config`.

**Comment by allevo** [comment]
> I think the two sections of the documentation should reference each other.

@jsumners I don't understand which sections are you'd like to see be linked

The documentation of plugins doesn't mention the route definition.

File: fastify.d.ts (hunk 3 above)
Comment (allevo) [reply fastify.d.ts]: Updated

**Comment by jsumners** [comment]
@allevo https://github.com/fastify/fastify/pull/884#issuecomment-381352986

**Comment by allevo** [comment]
> https://github.com/fastify/fastify/blob/5fefda9dd2f882e807b7ae736d39be4860c13d6b/docs/Plugins.md#plugin-options 

@jsumners  That was about plugin. This is for route definition

**Comment by jsumners** [comment]
They are the same thing. A route is just a plugin via an extra API method.

**Comment by allevo** [comment]
> They are the same thing. A route is just a plugin via an extra API method.

Sorry I don't follow you

```js
function handler (request, reply) {
  reply.send(reply.context.config)
}
fastify.register(function (f, o, n) {
  f.get('/inside', handler)
  n()
}, { config: { foo: 'bar' } })

fastify.get('/outside', handler)

fastify.inject({ method: 'GET', url: '/inside' })
  .then(response => {
    console.log(response.payload) // {"url":"/inside"}
  })

fastify.inject({ method: 'GET', url: '/outside' })
  .then(response => {
    console.log(response.payload) // {"foo":"bar","url":"/outside"}
  })
```

**Comment by jsumners** [comment]
From https://www.fastify.io/docs/latest/Getting-Started/

> As with JavaScript everything is an object, with Fastify everything is a plugin.

**Comment by mcollina** [comment]
@allevo @jsumners what's the status of this?

**Comment by jsumners** [comment]
I'm not blocking it.

**Pull request merged**